### PR TITLE
Feature/sidebar mode audit

### PR DESCRIPTION
**Pull request by piyalbasu**
This pull request introduces several improvements and bug fixes related to the extension's sidebar functionality, including more robust sidebar connection management, improved message routing, and better handling of signing request rejection. It also adds and updates tests to ensure correct behavior and compatibility across Chrome and Firefox.

**Sidebar connection and lifecycle management:**

* Added a new module (`sidebarPort.ts`) to manage the sidebar's long-lived port, allowing direct communication and safer handling of sidebar state transitions. [[1]](diffhunk://#diff-29e8d399a6422d87bcd209b6ea4e197d9165f9a6da0bd3dc9d56ed452b1c6cb3R1-R12) [[2]](diffhunk://#diff-1f6671130345a11d8901bfbcdfa2be09e82154b09435b4f177dd9c381ff690dcL55-R147)
* Improved sidebar disconnect logic to only reject signing requests handled by the sidebar, preventing accidental cancellation of requests managed by popups.
* Enhanced logic to differentiate between sidebar reloads and true disconnects, deferring cleanup to avoid prematurely rejecting requests.
* Introduced `sidebarQueueUuids` to track which signing requests are handled by the sidebar. [[1]](diffhunk://#diff-0752897b6ee1a578f2071f3540ef98755595a87cbcf5f8c890a4e9da8dca40beR18-R22) [[2]](diffhunk://#diff-1f6671130345a11d8901bfbcdfa2be09e82154b09435b4f177dd9c381ff690dcL55-R147)

**Message types and API changes:**

* Added new `REJECT_SIGNING_REQUEST` service type and message, and removed unused sidebar registration messages. [[1]](diffhunk://#diff-ed6526d491f387102ecf6556ba9725041c774be5c9004689f0fa34d045d4387bL66-R71) [[2]](diffhunk://#diff-0069fad9aa1015c453e2358e53b117d0307a27014f0f9c076fe923ee37955be4L448-R457) [[3]](diffhunk://#diff-0069fad9aa1015c453e2358e53b117d0307a27014f0f9c076fe923ee37955be4L527-R524)
* Added `rejectSigningRequest` API method for programmatically rejecting signing requests.

**Sidebar user experience and manifest:**

* Updated the manifest to add a `sidebar_action` entry for better Firefox support and improved sidebar icon handling.
* Clarified and improved logic for the "open sidebar by default" feature, including Chrome/Firefox compatibility notes.

**Message routing and navigation:**

* Changed sidebar navigation to use direct port messaging instead of broadcasting, improving reliability and security. [[1]](diffhunk://#diff-96f7a3637d5d67d1c18eb2b9f7c8312edd3eaa1706e0c0467f922626010c7271L27-R30) [[2]](diffhunk://#diff-96f7a3637d5d67d1c18eb2b9f7c8312edd3eaa1706e0c0467f922626010c7271L71-R84)

**Testing improvements:**

* Added and updated tests for sidebar behavior, settings, and port management to ensure robust handling of edge cases and cross-browser support. [[1]](diffhunk://#diff-9c6edc2b45c2685d0e00da6ee62534280bc8f089212ce46f39dcaca36b9d0f13R1-R120) [[2]](diffhunk://#diff-66507264a9272c79425785eee0f1446a95e99da6a33a4e2f7b31d0ba71459fe9R1-R171)

These changes collectively make the sidebar more robust, reliable, and compatible across browsers, while ensuring that signing requests and user actions are handled safely and predictably.

https://github.com/user-attachments/assets/8b52004d-03bc-4b5e-b345-de02e6f5de44

### CODE REVIEW

File: extension/src/background/messageListener/handlers/saveSettings.ts [hunk 1]
```diff
@@ -59,8 +59,7 @@ export const saveSettings = async ({
     isSorobanPublicEnabled: featureFlags.useSorobanPublic,
     isNonSSLEnabled: await getIsNonSSLEnabled({ localStore }),
     isHideDustEnabled: await getIsHideDustEnabled({ localStore }),
-    isOpenSidebarByDefault: (await localStore.getItem(
-      IS_OPEN_SIDEBAR_BY_DEFAULT_ID,
-    )) as boolean ?? false,
+    isOpenSidebarByDefault:
+      (await localStore.getItem(IS_OPEN_SIDEBAR_BY_DEFAULT_ID)) === "true",
```
[the remark this answers is not in the document]
Comment (piyalbasu) [reply extension/src/background/messageListener/handlers/saveSettings.ts]: Resolved in next commit

File: extension/src/background/messageListener/handlers/loadSettings.ts [hunk 2]
```diff
@@ -32,8 +32,7 @@ export const loadSettings = async ({
   const isNonSSLEnabled = await getIsNonSSLEnabled({ localStore });
   const isHideDustEnabled = await getIsHideDustEnabled({ localStore });
   const isOpenSidebarByDefault =
-    ((await localStore.getItem(IS_OPEN_SIDEBAR_BY_DEFAULT_ID)) as boolean) ??
-    false;
+    (await localStore.getItem(IS_OPEN_SIDEBAR_BY_DEFAULT_ID)) === "true";
   const { hiddenAssets } = await getHiddenAssets({ localStore });
```
[the remark this answers is not in the document]
Comment (piyalbasu) [reply extension/src/background/messageListener/handlers/loadSettings.ts:37]: Resolved in next commit

File: extension/src/background/index.ts [hunk 3]
```diff
@@ -135,17 +191,38 @@ export const initInstalledListener = () => {
...
   const val =
-    ((await localStore.getItem(IS_OPEN_SIDEBAR_BY_DEFAULT_ID)) as boolean) ??
-    false;
+    (await localStore.getItem(IS_OPEN_SIDEBAR_BY_DEFAULT_ID)) === "true";
   if (chrome.sidePanel?.setPanelBehavior) {
+    // Chrome: delegate action-click to the side panel when enabled
     chrome.sidePanel
-      .setPanelBehavior({ openPanelOnActionClick: !!val })
+      .setPanelBehavior({ openPanelOnActionClick: val })
       .catch((e) => console.error("Failed to set panel behavior:", e));
```
[the remark this answers is not in the document]
Comment (piyalbasu) [reply extension/src/background/index.ts:220]: Resolved in next commit

File: extension/src/background/messageListener/__tests__/loadSaveSettings.test.ts [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+};
+
+describe("loadSettings isOpenSidebarByDefault", () => {
+  it('returns true when localStorage value is the string "true"', async () => {
+    const localStore = {
+      getItem: jest.fn().mockImplementation((key: string) => {
+        if (key === IS_OPEN_SIDEBAR_BY_DEFAULT_ID)
+          return Promise.resolve("true");
+        return Promise.resolve(null);
+      }),
```
[the remark this answers is not in the document]
Comment (piyalbasu) [reply extension/src/background/messageListener/__tests__/loadSaveSettings.test.ts]: Resolved in next commit

File: extension/src/background/messageListener/freighterApiMessageListener.ts [hunk 5]
```diff
@@ -236,25 +259,33 @@ export const freighterApiMessageListener = (
...
+          safeResolve({
             apiError: FreighterApiInternalError,
           });
         } else if (popup !== null) {
           browser.windows.onRemoved.addListener(() =>
-            resolve({
+            safeResolve({
               apiError: FreighterApiDeclinedError,
             }),
           );
```
[the remark this answers is not in the document]
Comment (piyalbasu) [reply extension/src/background/messageListener/freighterApiMessageListener.ts]: @copilot apply changes based on [this feedback](https://github.com/stellar/freighter/pull/2674#discussion_r3023730932)

**Comment by piyalbasu** [comment]
@copilot apply changes based on the comments in [this thread](https://github.com/stellar/freighter/pull/2674#pullrequestreview-4052035893)

File: extension/src/background/index.ts [hunk 6]
```diff
@@ -52,19 +62,68 @@ export const initContentScriptMessageListener = () => {
...
+    // and reject any pending signing requests
     port.onDisconnect.addListener(() => {
+      clearSidebarPort();
       clearSidebarWindowId();
+
+      // Reject all active signing requests that were open in the sidebar
+      for (const uuid of activeQueueUuids) {
+        const responseIndex = responseQueue.findIndex(
+          (item) => item.uuid === uuid,
+        );
```
[the remark this answers is not in the document]
Comment (piyalbasu) [reply extension/src/background/index.ts]: @copilot apply changes based on [this feedback](https://github.com/stellar/freighter/pull/2674#discussion_r3029306959)

File: extension/src/background/index.ts [hunk 7]
```diff
@@ -52,19 +62,68 @@ export const initContentScriptMessageListener = () => {
...
+        setSidebarWindowId((msg as { windowId: number }).windowId);
       }
     });
 
     // When sidebar closes (for any reason), clear the window ID
+    // and reject any pending signing requests
     port.onDisconnect.addListener(() => {
+      clearSidebarPort();
       clearSidebarWindowId();
+
```
[the remark this answers is not in the document]
Comment (piyalbasu) [reply extension/src/background/index.ts:121]: @copilot apply changes based on [this feedback](https://github.com/stellar/freighter/pull/2674#discussion_r3029306995)

**Comment by piyalbasu** [comment]
### Code review

Found 3 issues:

1. `setAllowedStatus` is missing `sidebarQueueUuids` tracking, unlike every other signing handler

Every other handler (`requestAccess`, `submitToken`, `submitTransaction`, `submitBlob`, `submitAuthEntry`) adds its UUID to `sidebarQueueUuids` when routed to the sidebar and deletes it in `safeResolve`. `setAllowedStatus` was updated with the `safeResolve` pattern in this PR but omits both `sidebarQueueUuids.add(uuid)` and `sidebarQueueUuids.delete(uuid)`. If the sidebar disconnects while a `setAllowedStatus` request is pending, the disconnect handler in `index.ts` won't find the UUID in `sidebarQueueUuids`, so the dapp's promise hangs until the TTL timeout instead of being rejected immediately.

https://github.com/stellar/freighter/blob/26200d01927b38224c323f509d754e2b224dbc91/extension/src/background/messageListener/freighterApiMessageListener.ts#L783-L800

2. `ConfirmSidebarRequest` "Reject" button navigates home but does not reject the pending signing request

`handleReject` only calls `navigate(ROUTES.account)`. The new signing request that triggered this interstitial is still sitting in `responseQueue` with no resolution. The dapp is left with a hanging promise until the 5-minute TTL fires or the sidebar disconnects. All other rejection paths call `responseQueueItem.response(undefined)` to signal rejection back to the dapp immediately.

https://github.com/stellar/freighter/blob/26200d01927b38224c323f509d754e2b224dbc91/extension/src/popup/views/ConfirmSidebarRequest/index.tsx#L39-L42

3. `openSidebar()` in `navigate.ts` moves `window.close()` inside the `try` block, so the popup is never closed on error

Previously `window.close()` ran unconditionally after the `try/catch`. Now if `chrome.sidePanel.setOptions` or `.open` throws (e.g., API unavailable), execution jumps to `catch` and `window.close()` is skipped, leaving the popup orphaned.

https://github.com/stellar/freighter/blob/26200d01927b38224c323f509d754e2b224dbc91/extension/src/popup/helpers/navigate.ts#L30-L35

🤖 Generated with [Claude Code](https://claude.ai/code)

<sub>- If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

File: extension/src/popup/views/ConfirmSidebarRequest/index.tsx [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        const parsed = parsedSearchParam(queryString);
+        if (parsed.uuid) {
+          await internalRejectAccess({ uuid: parsed.uuid });
+        }
+      }
+    } catch {
+      // Best-effort — navigate home regardless
+    }
+    navigate(ROUTES.account);
+  };
```
[the remark this answers is not in the document]
Comment (piyalbasu) [reply extension/src/popup/views/ConfirmSidebarRequest/index.tsx:60]: @copilot apply changes based on [this feedback](https://github.com/stellar/freighter/pull/2674#discussion_r3030042973)

File: extension/src/background/messageListener/handlers/rejectSigningRequest.ts [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+  blobQueue: BlobQueue;
+  authEntryQueue: EntryQueue;
+  tokenQueue: TokenQueue;
+}) => {
+  const { uuid } = request;
+
+  if (!uuid) {
+    captureException("rejectSigningRequest: missing uuid in request");
+    return;
+  }
```
[the remark this answers is not in the document]
Comment (piyalbasu) [reply extension/src/background/messageListener/handlers/rejectSigningRequest.ts]: @copilot apply changes based on [this feedback](https://github.com/stellar/freighter/pull/2674#discussion_r3030424641)

File: extension/src/background/messageListener/popupMessageListener.ts [hunk 10]
```diff
@@ -27,7 +27,9 @@ import { publicKeySelector } from "background/ducks/session";
 import {
   startQueueCleanup,
   activeQueueUuids,
+  sidebarQueueUuids,
 } from "background/helpers/queueCleanup";
+import { getSidebarPort } from "./freighterApiMessageListener";
```
[the remark this answers is not in the document]
Comment (piyalbasu) [reply extension/src/background/messageListener/popupMessageListener.ts:32]: Fixed in next commit

File: extension/public/static/manifest/v3.json [hunk 11]
```diff
@@ -38,5 +38,13 @@
...
     "default_path": "index.html?mode=sidebar"
   },
+  "sidebar_action": {
+    "default_panel": "index.html?mode=sidebar",
+    "default_icon": {
+      "16": "images/icon16.png",
+      "32": "images/icon32.png"
+    },
+    "open_at_install": false
+  },
```
[the remark this answers is not in the document]
Comment (piyalbasu) [reply extension/public/static/manifest/v3.json:48]: Both browsers silently ignore the other's key, so having both in the same manifest is the correct cross-browser approach.

File: extension/src/background/messageListener/popupMessageListener.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (piyalbasu) [reply extension/src/background/messageListener/popupMessageListener.ts:32]: fixed in next commit

File: extension/src/background/index.ts [hunk 12]
```diff
@@ -52,19 +63,88 @@ export const initContentScriptMessageListener = () => {
...
+    // When sidebar closes, clear port state and schedule cleanup.
+    // The cleanup is deferred because chrome.sidePanel.open() can
+    // reload the sidebar page, causing a brief disconnect/reconnect.
     port.onDisconnect.addListener(() => {
-      clearSidebarWindowId();
+      if (getSidebarPort() === port) {
+        clearSidebarPort();
+        clearSidebarWindowId();
+      }
+
```
[the remark this answers is not in the document]
Comment (piyalbasu) [reply extension/src/background/index.ts]: Resolved in next commit

File: extension/src/background/messageListener/popupMessageListener.ts [hunk 13]
```diff
@@ -559,13 +580,18 @@ export const popupMessageListener = (
       const { uuid, isActive } = request as MarkQueueActiveMessage;
       if (isActive) {
         activeQueueUuids.add(uuid);
+        if (getSidebarPort() !== null) {
+          sidebarQueueUuids.add(uuid);
+        }
       } else {
```
[the remark this answers is not in the document]
Comment (piyalbasu) [reply extension/src/background/messageListener/popupMessageListener.ts:585]: Resolved in next commit

File: extension/specs/SIDEBAR_MODE.md [hunk 14]
```diff
@@ -0,0 +1,153 @@
+# Sidebar Mode Implementation Spec
```
Comment (piyalbasu) [inline extension/specs/SIDEBAR_MODE.md:1]: I added this spec because there ended up being some "gotchas" with this implementation.

First, there's the difference in browser implementation

Second, the inability to add e2e tests with Playwright

I asked Claude how we can help future agents have an easier time making changes and it guided me towards this solution: this type of spec that is linked in CLAUDE.MD should help make any architectural decisions and prevent agents from going down dead-ends we've already investigated in the future

Comment (piyalbasu) [reply extension/specs/SIDEBAR_MODE.md:1]: It's also nice to have a human readable description of what we're doing here for future human devs! 😄

**Comment by CassioMG** [comment]
@piyalbasu overall I think the PR is looking good! Posting here some findings and suggestions from superpowers for you to take a look - thanks!


### Important

#### I1. `REJECT_SIGNING_REQUEST` handler lacks `isFromExtensionPage` guard

**File:** `extension/src/background/messageListener/popupMessageListener.ts` — `REJECT_SIGNING_REQUEST` case (~line 355)

Unlike `OPEN_SIDEBAR` which explicitly checks `if (!isFromExtensionPage) return { error: "Unauthorized" }`, `REJECT_SIGNING_REQUEST` has no sender validation. This handler resolves a dApp's pending signing promise with `undefined` — defense-in-depth warrants an explicit guard.

**Fix:** Add `if (!isFromExtensionPage) return { error: "Unauthorized" }` at the top of the case.

#### I2. Duplicated queue cleanup logic

**Files:** `extension/src/background/index.ts` (disconnect handler, ~lines 121-145) and `extension/src/background/messageListener/handlers/rejectSigningRequest.ts`

The disconnect handler duplicates the exact same iterate-find-splice-resolve pattern across all 5 queues. If either is updated (e.g., a new queue type), the other must be updated in lockstep.

**Fix:** Have the disconnect handler call `rejectSigningRequest` for each UUID, or extract a shared `removeFromAllQueues(uuid, queues)` helper.

#### I3. `ConfirmSidebarRequest` uses raw HTML instead of design system components

**File:** `extension/src/popup/views/ConfirmSidebarRequest/index.tsx` — lines 68-76

Uses `<h1>` with custom CSS font-size/font-weight and `<div>` for body text. Per project conventions: "Always use `Text` or `Heading` from `@stellar/design-system` — never raw HTML text elements with custom font styles."

**Fix:**
```tsx
// Replace:
<h1 className="ConfirmSidebarRequest__title">{t("New Signing Request")}</h1>
<div className="ConfirmSidebarRequest__body">{t("A new signing request...")}</div>

// With:
<Heading as="h1" size="md" addlClassName="ConfirmSidebarRequest__title">{t("New Signing Request")}</Heading>
<Text as="p" size="sm" addlClassName="ConfirmSidebarRequest__body">{t("A new signing request...")}</Text>
```

Also in `styles.scss`: replace `font-weight: 500` with `var(--font-weight-medium)`, `font-weight: 400` with `var(--font-weight-regular)`, and remove the hardcoded `font-family: "Inter", sans-serif`.

#### I4. `requestAccess` and `setAllowedStatus` missing `rejectOnWindowClose` for popup windows (pre-existing)

**File:** `extension/src/background/messageListener/freighterApiMessageListener.ts`

`requestAccess` (~line 174) has `if (popup === null)` (sidebar TTL fallback) but no `else if (popup !== null)` branch calling `rejectOnWindowClose`. If the user closes the popup window, the dApp promise hangs until TTL (~5 min). Same for `setAllowedStatus` (~line 770).

Compare with `submitTransaction`, `submitBlob`, `submitAuthEntry` which all have the three-way branch: `undefined` (error), `null` (sidebar TTL), `!== null` (popup window close via `rejectOnWindowClose`).

**Note:** Pre-existing — not introduced by this PR. But since this audit added `rejectOnWindowClose` as a consolidated helper, it's a good opportunity to fix these two.

**Fix:** Add `else if (popup !== null) { rejectOnWindowClose(popup.id!, safeResolve); }` to both flows.

#### I5. `submitToken` missing sidebar TTL fallback (pre-existing)

**File:** `extension/src/background/messageListener/freighterApiMessageListener.ts` — submitToken (~line 283)

Has `popup === undefined` (error) and `popup !== null` (window close via `rejectOnWindowClose`), but no `popup === null` (sidebar) branch with TTL fallback. Inconsistent with the other signing flows. The sidebar disconnect cleanup via `sidebarQueueUuids` covers this, but the explicit TTL fallback is a safety net.

**Note:** Pre-existing.

---

## Code Quality

#### C1. `import` statement placed after function definition

**File:** `extension/src/background/messageListener/freighterApiMessageListener.ts` — ~line 126

```ts
const rejectOnWindowClose = (...) => { ... };

import { DataStorageAccess } from "background/helpers/dataStorageAccess";
```

Import placed between the `rejectOnWindowClose` function and the main export. Works due to hoisting, but imports should be at the top of the file.

#### C2. `safeResolve` typed as `(value: any) => void`

**File:** `extension/src/background/messageListener/freighterApiMessageListener.ts` — `rejectOnWindowClose` and all `safeResolve` closures

Using `any` loses type safety on the resolve value. Could be typed generically or at minimum use `unknown`.

#### C3. `SIDEBAR_NAVIGATE` exported as plain string constant, not in an enum

**File:** `@shared/constants/services.ts:77`

All other service types use the `SERVICE_TYPES` or `EXTERNAL_SERVICE_TYPES` enums. A comment explaining why it's separate (only used on ports, not through standard message routing) would help.

#### C4. `safeResolve` duplicated 6 times inline

**File:** `extension/src/background/messageListener/freighterApiMessageListener.ts`

The identical closure appears in all 6 signing handlers. Consider extracting a `makeSafeResolve` helper:

```ts
const makeSafeResolve = <T>(resolve: (value: T) => void) => {
  let resolved = false;
  return (value: T) => {
    if (resolved) return;
    resolved = true;
    resolve(value);
  };
};
```

---

## Suggestions

#### S1. Derive `SIGNING_ROUTE_PREFIXES` from `ALLOWED_NAV_PREFIXES`

**File:** `extension/src/popup/components/SidebarSigningListener/index.tsx`

Two arrays that must stay in sync. Derive one from the other:

```ts
const SIGNING_ROUTE_PREFIXES = [...ALLOWED_NAV_PREFIXES, ROUTES.confirmSidebarRequest];
```

#### S2. Extract 500ms debounce to a named constant

**File:** `extension/src/background/index.ts` — ~line 147

```ts
export const SIDEBAR_DISCONNECT_DEBOUNCE_MS = 500;
```

#### S3. Guard `ConfirmSidebarRequest` route against non-sidebar access

**File:** `extension/src/popup/Router.tsx`

The route is registered globally. Consider redirecting to `ROUTES.account` if `!isSidebarMode()`.

#### S4. `window.close()` in `navigate.ts` — use `finally`

**File:** `extension/src/popup/helpers/navigate.ts`

`window.close()` now appears in both try and catch. A `finally` block would be clearer.

---

## Missing Test Coverage

1. **`rejectSigningRequest` handler** — No dedicated tests. Should verify: UUID found in responseQueue -> response called with `undefined`; UUID not found -> Sentry exception captured; missing UUID -> early return; all queue types cleaned up.

2. **`ConfirmSidebarRequest` component** — `isValidNextRoute` contains security-sensitive open-redirect validation. Should test edge cases: `//evil.com`, `javascript:alert(1)`, empty strings, encoded characters, valid routes.

3. **`SidebarSigningListener` route allowlist** — No tests verifying disallowed routes are blocked.

---

## Nits

| Location | Issue |
|----------|-------|
| `ConfirmSidebarRequest/styles.scss:27` | `--sds-clr-amber-*` tokens — verify these exist in the SDS token set (CLAUDE.md only documents grayscale and red) |
| `ConfirmSidebarRequest/styles.scss:37` | `font-weight: 500` → use `var(--font-weight-medium)` |
| `ConfirmSidebarRequest/styles.scss:45` | `font-weight: 400` → use `var(--font-weight-regular)` |
| `queueCleanup.ts:18-22` | Comment could link to the actual cleanup code location |

---

## Summary Table

| Category | Count | Items |
|----------|-------|-------|
| Important | 5 | I1-I5 (I4-I5 pre-existing) |
| Code Quality | 4 | C1-C4 |
| Suggestions | 4 | S1-S4 |
| Missing Tests | 3 | rejectSigningRequest, ConfirmSidebarRequest, route allowlist |
| Nits | 4 | Style conventions |
